man/dblgen.1
.TH DBLGEN 1 "12 September 1995"
.FB
.SH NAME
dblgen \- generate a set of \fIFirstBase\fP labels
.SH SYNOPSIS
.B dblgen
[\fB-b\fP] [\fB-y\fP] [\fB-n\fP]
[\fB-d\fP \fIdatabase\fP] [\fB-i\fP \fIindex\fP]
[\fB-H\fP \fIN_lines\fP]
[\fB-T\fP \fIN_lines\fP]
.SH DESCRIPTION
.I Dblgen
generates a set of \fIFirstBase\fP labels using
a \fIFirstBase\fP label dictionary defined via 
\fIdbdlbl\fP(1).
This label dictionary is used to determine which fields are to appear
where on the label.
.PP
The \fB-b\fP
flag will run the \fIFirstBase\fP tool in batch mode, suitable for shell
scripts.
This mode produces no output on the terminal, thus can also be used
when running the process in the background.
The \fB-y\fP flag will run the tool at the terminal without pausing to ask
the user if existing files can be destroyed and without displaying
a 'picture' of the requested task.
.PP
\fIDblgen\fP(1)
first displays an image of what the labels will look like and where the
print image file will go. The user has a chance to abort here.
The labels may be generated from one to four up (across the top), as
specified in the label dictionary. The width of each label is
derived from the label dictionary also. See
\fIdbdlbl\fP(1)
for more information on the label generator dictionary.
.PP
The \fB-n\fP argument line switch is used to turn off the printing of the
statistics label. Without this flag, a statistics label showing how many
labels were printed is produced as the last label.
.PP
The \fB-H\fP and \fB-T\fP
argument line switches are used to tune the number of blanklines
printed \fIbetween\fP each label, either as head_lines or tail_lines.
Using this feature, printer fonts
requiring more vertical spacing can be utilized as needed.
.SH FILES
.PD 0
.TP 10
cdblbl
default name for the \fIFirstBase\fP labels.
.TP 10
dbase
default name for the \fIFirstBase\fP database.
.TP 10
index
default name for the \fIFirstBase\fP index.
.TP 10
*.idictl
dictionary for label generator.
.TP 10
*.cdb
\fIFirstBase\fP database.
.TP 10
*.idx
\fIFirstBase\fP index.
.TP 10
*.lbl
\fIFirstBase\fP labels.
.PD
.SH SEE ALSO
dbedit(1), dbdlbl(1), generators(5), firstbase(5)
.PP
.I FirstBase User's Guide and Reference Manual
.br
.SH DIAGNOSTICS
If the label print file already exists 
\fIDblgen\fP(1)
will pause and ask permission to overwrite it.
.sp 1
This tool uses a \fIFirstBase\fP index to generate the output from the database
in a particular order.
If this index is detected to be out of date or non-existent, it is
regenerated \fIautomatically\fP.
.br


ChangeLog
2012-04-14    <john.bolding@epona.horseweb.com>

	* modified libscr/input.c -- read_kcheck -- to avoid a problem when:
	     EDITINPUT is OFF
	     two (2) ESCAPE's were hanging in read_kcheck

2011-09-29    <john.bolding@epona.horseweb.com>

	* added fflush(-1) to dbmacro via libmexpr/mac_file.c

2009-02-25    <john.bolding@epona.horseweb.com>

	* changed src/dbsql/tokenscript to allow for space between # and define.

2006-09-19  John  <john@cerebus.horseweb.com>

	* complete rewrite of libcc/subline.c to NOT write on the passed in s.

2003-03-29  John Bolding  <john@firstbase.com>

        * tweaked libmpars/tokenscript and dbsql/tokenscript
          to accomodate the bison output "# define" instead of "#define".

2001-09-01  John Bolding  <john@firstbase.com>

        * tweaked configure.in to not use termios.h if OS is sco3.2
	* src/libmexpr/mac_func.c
	* src/libmexpr/mac_str.c
	  augmented dbmacro to include
	     crypt(string, salt), returns an encrypted string ala crypt(3)

2001-03-19  John Bolding  <john@firstbase.com>

	* src/libmexpr/mac_sec.c: 
	* src/libmexpr/mac_func.c: 
	* src/libmexpr/mac_file.c: 
	* src/include/macro.h:
	  modified dbmacro functions as follows:
	  - modified mf_creat() to do a chmod on the file if it already exists.
	  - added an mf_mkstemp() that really calls mf_mktemp()
	    (since mf_mktemp really uses mkstemp(2), it closes the fd.
	     a real mf_mkstemp would modify the array passed in for
	     the user to keep. that did not happen this go around)
	  - renamed all the firstbase secure record functions to 
            r_owner() r_group() r_mode() r_chown() r_chgrp() r_chmod()
	  - added an mf_chmod() command

2001-03-18  John Bolding  <john@firstbase.com>

	* src/dbtools/dbmacro.c: fixed path_trans to be null before
	  its used. fixed to only look at extension of file name, .cgi,
	  to determine whether to go cgi mode. this is so that i can
	  test with program names like ndbmacro.cgi.

	* src/libcdb/cgi.c: modified cgi_read_get to FB_ERROR out
	  if the QUERY_STRING env variable is not set.

	* src/libcdb/System.c: seems linux system() call is a bit
	  flaky at times on some versions, so fb_system() has been
	  modified to not use the system() call at all.
	  now it uses execvp. in addition, changed to waitpid()
	  to try and get better response from cgi dbmacro scripts that
	  use the macro call system(). and, now pays attention to
	  status from waitpid() and exits if its strange.

2001-02-23  John Bolding  <john@firstbase.com>

	* src/libscr/screen.c: modified visible bell area to
	  use a ^G if its not defined.
	  this was a terminfo only bug --- termcap was already working.

2001-02-17  John Bolding  <john@firstbase.com>

	* release 9.3.1
	* man pages for section 4 were missing from distribution.
	  (finally noticed by <garak_dracul@hotmail.com>).

2001-02-16  John Bolding  <john@firstbase.com>

	* man.ps target now in man/Makefile

	* fixed use of mktemp and gets system wide

	* src/libinit/setup.c:
	  changed default behaviour. these are now default:
	     MENUPAUSE OFF
	     MAKEINDEX OFF
	     DECIMAL ON

	* quite a few bugs regarding #if clauses and FSYNC/FCNTL
	  these should have been HAVE_FSYNC or HAVE_FCNTL.
	  some were wrong. some were using FSYNC instead of HAVE_FSYNC.
	  this might have caused multiuser problems with records
	  stored across slow devices (NFS, etc).
	  all of these were introduced when firstbase went autoconf.
	
	* src/libcdb/System.c:
	  - added \" to list of characters that cause a complete system() call
	  - make return status better for those that fork and wait

	* a number of little bugs concering the full screen display
	  for devices bigger than 24 x 80.
	
	* 180 degree about face regarding TERMCAP vs TERMINFO:
	  since ncurses seems to be pretty common on Linux/FreeBSD/Etc,
	  the default is now to use the ncurses library and if this
	  is not found the HAVE_LIBNCURSES is undef'ed in
	  src/include/config.h.

	  Although I am not a TERMINFO fan, perhaps this will be good.
	  At some point, the firstbase screen libraries could
	  make better use of ncurses to optomize screen motion, etc.
	  Perhaps even support for colors, pop-up style ncurses items, etc.

	  So, TERMINFO via ncurses is the default.
	  And, TERMCAP is done only if necessary.

2001-02-05  John Bolding  <john@firstbase.com>

	* src/libscr/input.c: 
	* src/libscr/e_input.c: both of these were modified to handle
	  windows that change sizes while being used, ala SIGWINCH.

	* src/libscr/refresh.c: 
	* src/libscr/scrhdr.c: 
	* src/libscr/scrstat.c: 
	* src/libscr/scrtime.c: 
	* src/libinit/ender.c: 
	* src/libcdb/screrr.c:
	  all of these were modified to better handle standard
	  screen messages on surfaces bigger than 24 x 80,
	  and to restore headers and footers when the frame size changes.

	* src/libcdb/winsize.c: modified to handle different frame/window
	  sizes using an ioctl() call with TIOCGWINSZ or TIOCGSIZE,
	  then allowing environment variables LINES and COLUMNS to
	  override all else. (Thanks to <vanonim@dial.eunet.ch>).

	  finally, sanity checks are done to make sure
	  nothing interactive less than 24 x 80 gets through.

	* src/libinit/getargs.c: took out code that was getting in the way
	  of the new fb_winsize(). all sanity checks are now done at
	  screen setup time.

	* src/libscr/screen.c: force all columns and lines determinations
	  to be done inside of winsize, newly named fb_winsize().

2001-02-02  John Bolding  <john@firstbase.com>

	* src/libinit/setup.c: removed the LINES setup variable,
	     (also force makeindex to be OFF at startup).

2001-02-02  John Bolding  <john@firstbase.com>

	* .visual: changed to /usr/bin/vi (was /usr/ucb/vi).

	* configure.in (INCLUDES): made `ed' and one of `flex' or `lex'
	  and one of `bison', `byacc', or `yacc' required in order
	  to complete ./configure.

$Id: ChangeLog,v 9.9 2006/09/19 17:46:07 john Exp $


src/dblgen/dblgen.c
/*
 * Copyright (C) 2001 FirstBase Software, Inc. All rights reserved.
 *
 * $Id: dblgen.c,v 9.1 2001/02/16 19:43:22 john Exp $
 *
 * See the file LICENSE for conditions of use and distribution.
 *
 */

#ifndef LINT
static char sccsid[] = "@(#)dblgen.c	8.3 11/26/93 FB";
#endif

/*
 *  dblgen - generate labels (uses idictl)
 */

#include <fb.h>
#include <fb_vars.h>

#define LABEL_SIZE 55
#define MAXLABEL 250

extern short int cdb_interrupt;
extern short int cdb_screenprint;
extern short int cdb_secure;

static int curlbl = 0;			/* to mark current label point */
static int justify = 0;			/* set to right justify header line */
static int asize = 0;			/* default label size */

static lgen();
static maklab();
static void dump();
static one_lgen();
static usage();

/*
 *  surely there will not be more that FB_MAXKEEP (1000) fields used
 *  for a label. lets hope not, anyway...
 *
 *  keyprint is the list of fields in the label.
 *     each line has a label pointer slot of NULL to terminate the
 *     list of fields for that line. there may be a max of 5
 *     such lines.
 */
 
fb_field *keyprint[FB_MAXKEEP] = { NULL };
fb_field **pp;
fb_database *hp;

int nostat_label = 0;
int headlines = 0;
int taillines = 0;

char *pline[5], crec[FB_RECORDPTR+1], fld[MAXLABEL];
long rec, n, atol();
FILE *fs, *initlabel();
int fmt;

/*
 *  dblgen - main driver
 */
 
   main(argc, argv)
      int argc;
      char *argv[];
   
      {
         char filen[FB_MAXNAME];
         int p;

	 asize = LABEL_SIZE;
         fb_getargs(argc, argv, FB_ALLOCDB);
	 if (fb_testargs(argc, argv, "-n"))
	    nostat_label = 1;
	 if ((p = fb_testargs(argc, argv, "-H")) > 0)
	    headlines = atoi(argv[p + 1]);
	 if ((p = fb_testargs(argc, argv, "-T")) > 0)
	    taillines = atoi(argv[p + 1]);
	 hp = cdb_db;
	 fb_opendb(cdb_db, READ, FB_WITHINDEX, FB_MUST_INDEX);
         fb_scrhdr(cdb_db, "Parsing"); fb_infoline();
         fs = initlabel(cdb_keymap, keyprint, hp, 
                        filen, argv, hp->idict, &fmt, &asize);
	 if (fs == NULL){
	    sprintf(filen, "%sl", hp->idict);
	    fb_xerror(FB_BAD_DICT, filen, NIL);
	    }
	 if (!cdb_batchmode && !cdb_yesmode){
            ltrace(keyprint, filen, fmt, asize);
            if (fb_mustbe('y',
	          "If accurate, enter 'y' to continue:",
                  cdb_t_lines, 1) == FB_ERROR)
               fb_xerror(FB_ABORT_ERROR, SYSMSG[S_NOT_DONE], NIL);
	    fb_move(4, 1), fb_clrtobot(), fb_infoline();
	    }
	 if (fmt < 0){
	    fmt = -fmt;
	    justify = 1;
	    }
	 if (fmt * asize >= MAXLABEL)
	    fb_xerror(FB_MESSAGE, SYSMSG[S_TOO_LONG], NIL);
	 if (cdb_interrupt)
	    fb_allow_int();
         pp = keyprint;
         if (lgen() == FB_ERROR)
            fb_xerror(FB_ABORT_ERROR, SYSMSG[S_NOT_DONE], NIL);
	 fb_closedb(cdb_db);
	 fclose(fs);
	 if (!cdb_batchmode && cdb_screenprint)
	    fb_screenprint(filen);
         fb_ender();
      }
                  
/* 
 *  lgen - generate labels according to list in pp 
 */
 
   static lgen()

      {
         int i, one_lgen();

         n = 0L;
	 if (!cdb_batchmode){
            fb_scrstat("Printing");
	    FB_XWAIT();
	    fb_gcounter(n);
	    }
	 if (cdb_yesmode)
	    sleep(1);				/* some are too fast */
	 for (i = 0; i < 5; i++){
	    pline[i] = (char *) fb_malloc(MAXLABEL+1);
	    pline[i][0] = NULL;
	    }
         fb_forxeach(hp, one_lgen);
	 if (nostat_label == 0){
	    sprintf(fld, " %ld labels printed", n);
	    strcat(pline[0], fld);
	    curlbl++;
	    }
	 if (curlbl > 0){
	    curlbl = fmt - 1;	/* force dump */
	    dump(pline, fmt, fs);
	    }
	 return(FB_AOK);
      }

   static one_lgen(hp)
      fb_database *hp;

      {
         maklab(pline, pp);
         dump(pline, fmt, fs);
         ++n;
         if (!cdb_batchmode)
            fb_gcounter(n);
         return(FB_AOK);
      }

/* 
 *  maklab - make a label. fb_put into plines 
 */
 
   static maklab(pline, pp)
      char *pline[];
      fb_field *pp[];

      {
         char fld[MAXLABEL], pfld[MAXLABEL], *fb_trim(), tmp[MAXLABEL];
	 int i, p, j, head;

         strcpy(pfld, " ");
	 if (pp[0] == NULL)
	    head = 0;
	 else
	    head = 1;
	 for (p = 0, i = 0; ; i++){
	    if (pp[i] == NULL){
	       if (head == 1 && p == 0 && justify == 1){ /* justify header  */
		  sprintf(fld, "%100s", " ");
		  j = asize - strlen(pfld) - 1;
		  if (j < 0)
		     j = 0;
		  fld[j] = NULL;
		  strcat(fld, pfld);
		  }
	       else{				/* normal case */
	          fb_rmlead(pfld);
		  sprintf(fld, "%-100s", pfld);
	          fb_rmlead(fld);
		  fld[asize] = NULL;
		  }
	       strcpy(pfld, fld);
	       if (strlen(pfld) > 0)
		  strcat(pline[p++], pfld); 
	       pfld[0] = fld[0] = NULL;
	       continue;
	       }
	    else if (pp[i]->size == 0)
	       break;
	    if (pp[i]->size > 0){
	       fb_fetch(pp[i], fld, hp);
	       fb_formfield(tmp, fld, pp[i]->type, pp[i]->size);
	       fb_rmlead(fb_trim(tmp));
	       strcpy(fld, tmp);
	       }
	    else
	       strcpy(fld, pp[i]->id);
	    strcat(pfld, fld);
	    }
	 sprintf(fld, "%-100s", " ");
	 fld[asize] = NULL;
	 for (; p < 5; p++)
	    strcat(pline[p], fld);
	 }

/* 
 *  dump - dump plines is fmt labels already made 
 */
 
   static void dump(pline, fmt, fs)
      char *pline[];
      int fmt;
      FILE *fs;

      {
	 char *fb_trim();
	 int i;

	 if (++curlbl < fmt)
	    return;
	 for (i = 0; i < headlines; i++)
	    fprintf(fs, "\n");
	 for (i = 0; i < 5; i++){
	    fprintf(fs, "%s\n", fb_trim(pline[i]));
	    pline[i][0] = NULL;
	    }
	 for (i = 0; i < taillines; i++)
	    fprintf(fs, "\n");
	 curlbl = 0;
	 fprintf(fs, "\n");
      }

/* 
 *  usage message 
 */
 
   static usage()
      {
         fb_xerror(FB_MESSAGE,
            "usage: dblgen [-d dbase] [-i index] [-b] [-y]", NIL);
      }
